presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
+        partitionStatistics.forEach((partition, statistics) -> {
+            HiveBasicStatistics basicStatistics = statistics.getBasicStatistics();
+            OptionalLong rowCount = basicStatistics.getRowCount();
+            rowCount.ifPresent(count -> checkStatistics(count >= 0, table, partition, "rowCount must be greater than or equal to zero: %s", count));
+            basicStatistics.getFileCount().ifPresent(count -> checkStatistics(count >= 0, table, partition, "fileCount must be greater than or equal to zero: %s", count));
+            basicStatistics.getInMemoryDataSizeInBytes().ifPresent(size -> checkStatistics(size >= 0, table, partition, "inMemoryDataSizeInBytes must be greater than or equal to zero: %s", size));
+            basicStatistics.getOnDiskDataSizeInBytes().ifPresent(size -> checkStatistics(size >= 0, table, partition, "onDiskDataSizeInBytes must be greater than or equal to zero: %s", size));
+            statistics.getColumnStatistics().forEach((column, columnStatistics) -> validateColumnStatistics(table, partition, column, rowCount, columnStatistics));
+        });
     }
 
-    public static com.facebook.presto.spi.statistics.Range createPrestoRange(Type type, Object min, Object max)
+    private static void validateColumnStatistics(SchemaTableName table, String partition, String column, OptionalLong rowCount, HiveColumnStatistics columnStatistics)
     {
-        return new com.facebook.presto.spi.statistics.Range(convertPrestoTypeValueToDouble(type, min), convertPrestoTypeValueToDouble(type, max));
+        columnStatistics.getMaxValueSizeInBytes().ifPresent(maxValueSizeInBytes ->
+                checkStatistics(maxValueSizeInBytes >= 0, table, partition, column, "maxValueSizeInBytes must be greater than or equal to zero: %s", maxValueSizeInBytes));
+        columnStatistics.getTotalSizeInBytes().ifPresent(totalSizeInBytes ->
+                checkStatistics(totalSizeInBytes >= 0, table, partition, column, "totalSizeInBytes must be greater than or equal to zero: %s", totalSizeInBytes));
+        columnStatistics.getNullsCount().ifPresent(nullsCount -> {
+            checkStatistics(nullsCount >= 0, table, partition, column, "nullsCount must be greater than or equal to zero: %s", nullsCount);
+            if (rowCount.isPresent()) {
+                checkStatistics(
+                        nullsCount <= rowCount.getAsLong(),
+                        table,
+                        partition,
+                        column,
+                        "nullsCount must be less than or equal to rowCount. nullsCount: %s. rowCount: %s.",
+                        nullsCount,
+                        rowCount.getAsLong());
+            }
+        });
+        columnStatistics.getDistinctValuesCount().ifPresent(distinctValuesCount ->
+                checkStatistics(distinctValuesCount >= 0, table, partition, column, "distinctValuesCount must be greater than or equal to zero: %s", distinctValuesCount));
```

Comment:
Would you comment on why we can't check that `distinctValuesCount <= rowCount` (e.g. approximate value may be a bit larger?)

It might be worth renaming `distinctValuesCount` to reflect that the value is approximate, e.g. `approxDistinctValuesCount`

Reply:
I'm adding this check in the very next commit.